server/src/test/java/org/elasticsearch/action/ingest/SimulatePipelineResponseTests.java

```diff
@@ -120,4 +92,83 @@ public void testSerialization() throws IOException {
+            } else {
+                results.add(new SimulateDocumentBaseResult(ingestDocument));
+                SimulateDocumentBaseResult simulateDocumentBaseResult;
+                if (isFailure) {
+                    simulateDocumentBaseResult = new SimulateDocumentBaseResult(new IllegalArgumentException("test"));
+                } else {
+                    simulateDocumentBaseResult = new SimulateDocumentBaseResult(ingestDocument);
+                }
+                results.add(simulateDocumentBaseResult);
+            }
+        }
+        return new SimulatePipelineResponse(pipelineId, isVerbose, results);
+    }
+
+    @Override
+    protected SimulatePipelineResponse createTestInstance() {
+        boolean isVerbose = randomBoolean();
+        // since the pipeline id is not serialized with XContent we set it to null for equality tests.
+        return createInstance(null, isVerbose, false);
+    }
+
+    @Override
+    protected SimulatePipelineResponse doParseInstance(XContentParser parser) {
+        return SimulatePipelineResponse.fromXContent(parser);
+    }
+
+    @Override
+    protected boolean supportsUnknownFields() {
+        return false;
+    }
+
+    @Override
+    protected void assertEqualInstances(SimulatePipelineResponse response,
+                                        SimulatePipelineResponse parsedResponse) {
+        assertEquals(response.getPipelineId(), parsedResponse.getPipelineId());
+        assertEquals(response.isVerbose(), parsedResponse.isVerbose());
+        assertEquals(response.getResults().size(), parsedResponse.getResults().size());
+        for (int i=0; i < response.getResults().size(); i++) {
+            if (response.isVerbose()) {
+                assert response.getResults().get(i) instanceof SimulateDocumentVerboseResult;
+                assert parsedResponse.getResults().get(i) instanceof SimulateDocumentVerboseResult;
+                SimulateDocumentVerboseResult responseResult = (SimulateDocumentVerboseResult)response.getResults().get(i);
+                SimulateDocumentVerboseResult parsedResult = (SimulateDocumentVerboseResult)parsedResponse.getResults().get(i);
+                SimulateDocumentVerboseResultTests.assertEqualDocuments(responseResult, parsedResult);
+            } else {
+                assert response.getResults().get(i) instanceof SimulateDocumentBaseResult;
+                assert parsedResponse.getResults().get(i) instanceof SimulateDocumentBaseResult;
+                SimulateDocumentBaseResult responseResult = (SimulateDocumentBaseResult)response.getResults().get(i);
+                SimulateDocumentBaseResult parsedResult = (SimulateDocumentBaseResult)parsedResponse.getResults().get(i);
+                SimulateDocumentBaseResultTests.assertEqualDocs(responseResult, parsedResult);
+            }
+        }
+    }
+
+    @Override
+    protected boolean assertToXContentEquivalence() {
+        return false;
```

Comment:
is this because of an exception somewhere?

Reply:
Yes. If there are failures we have serialized `ElasticsearchException`s inside.

Reply:
could we separate normal testing (without exception) from testing with exceptions? We do something similar in ListTasksResponseTests

Reply:
This was actually good. I caught a bug. Thanks!